source.go

```diff
@@ -1,60 +1,145 @@
-	config, err := ParseSourceConfig(cfg)
+	sdk.Logger(ctx).Info().Msg("Configuring Weather Source Connector...")
+	var config SourceConfig
+	err := sdk.Util.ParseConfig(cfg, &config)
 	if err != nil {
 		return err
 	}
-
 	s.config = config
 	return nil
 }
 
 func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
+	s.client = &http.Client{}
+	s.url = s.CreateRequestURL()
+	// try pinging the URL with APPID
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, s.url, nil)
+	if err != nil {
+		return fmt.Errorf("error creating HTTP request %q: %w", s.config.URL, err)
+	}
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error pinging URL %q: %w", s.config.URL, err)
+	}
+	if resp.StatusCode == http.StatusUnauthorized {
+		return fmt.Errorf("authorization failed, check your APPID key")
+	}
+	defer resp.Body.Close()
+
+	s.limiter = rate.NewLimiter(rate.Every(s.config.PollingPeriod), 1)
+
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return sdk.Record{}, nil
+	if s.limiter.Allow() {
+		rec, err := s.getRecord(ctx)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("error getting the weather data: %w", err)
+		}
+		return rec, nil
+	}
+	return sdk.Record{}, sdk.ErrBackoffRetry
 }
 
 func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
-	return nil
+	sdk.Logger(ctx).Debug().Str("position", string(position)).Msg("got ack")
+	return nil // no ack needed
 }
 
 func (s *Source) Teardown(ctx context.Context) error {
+	if s.client != nil {
+		s.client.CloseIdleConnections()
+	}
 	return nil
 }
+
+func (s *Source) CreateRequestURL() string {
+	return s.config.URL + "?" + "q=" + s.config.City + "&" + "APPID=" + s.config.APPID + "&" + "units=" + s.config.Units
+}
+
+func (s *Source) getRecord(ctx context.Context) (sdk.Record, error) {
+	// create GET request
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.url, nil)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	// get response
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error getting data from URL: %w", err)
+	}
+	defer resp.Body.Close()
+	// read body
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error reading body for response %v: %w", resp, err)
+	}
+	// parse json
+	var structData sdk.StructuredData
+	err = json.Unmarshal(body, &structData)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed to unmarshal body as JSON: %w", err)
+	}
+	// create record
+	now := time.Now().Unix()
+	timestamp, ok := structData["dt"]
+	if !ok {
+		return sdk.Record{}, fmt.Errorf("dt field not found in record: %w", err)
+	}
+	rec := sdk.Record{
+		Payload: sdk.Change{
+			Before: nil,
+			After:  structData,
+		},
+		Operation: sdk.OperationCreate,
+		Position:  sdk.Position(fmt.Sprintf("%v", timestamp)),
+		Key:       sdk.RawData(fmt.Sprintf("unix-%v", now)),
```

Comment:
I assume that `dt` is the time when the weather was last updated, right? Then we should switch the position and key - records with the same key represent the same "entity" (e.g. weather measurement), the position is there to uniquely identify a record.

Reply:
oh I see! will switch them then